### PR TITLE
[refactor] estimate model FLOPs per input batch 

### LINKED ISSUE #4009: VLM MFU is miscomputed: vision-encoder FLOPs billed per decoder token, not per image patch
### Summary

The MoE/dense FLOP estimators in `torchtitan/models/utils.py` (`get_moe_model_nparams_and_flops`, `get_dense_model_nparams_and_flops`) do not model the vision encoder correctly for VLMs. The full VLM (decoder + `vision_encoder`) is passed in, but the vision-encoder parameters are not treated specially -- they fall into the catch-all `else` branch and are added to `nparams_dense`:

```python
for name, p in model.named_parameters():
    if "embedding" in name:
        nparams_embedding += p.numel()
        nparams_dense += p.numel()
    elif "moe.shared_experts" in name:
        ...
    else:
        nparams_dense += p.numel()      # <-- ALL vision_encoder params land here
...
nparams_for_matmul = nparams_dense - nparams_embedding + nparams_sparse_active   # includes Nv
num_flops_per_token = 6 * nparams_for_matmul + 6 * num_full_attn * n_heads * head_dims * seq_len
```

So the vision encoder is billed at `~6 * Nv` FLOPs **per decoder token**. But the vision encoder does not run per decoder token -- it runs once per **image patch**, and the number of patches depends on the images in the batch (`grid_thw`), not the decoder sequence length. The estimate therefore mis-scales, and since MFU divides these FLOPs by real step time, **reported VLM MFU is wrong** -- overstated for text-heavy batches, understated for image-heavy ones.

Surfaced in review of the Kimi-K2.5 PR (#3532).

### Who is affected

Any VLM that routes its full module (including a vision encoder) through these estimators. On `main` that is **`qwen3_5`** (the entire ladder -- every rung has a vision encoder). Text-only models are unaffected (no `vision_encoder` params).

### Quantifying the error (real qwen3_5 ladder)

Let `A` = active decoder params, `Nv` = vision-encoder params, `f` = fraction of the decoder sequence that is vision tokens. With `spatial_merge_size = 2` the encoder processes `merge^2 = 4` patches per decoder vision token, so real vision FLOPs `~= 6 * Nv * (4 * f * T)` while the code counts `6 * Nv * T`. The reported/true MFU ratio is:

```
reported / true = (1 + Nv/A) / (1 + r*f),    r = 4 * Nv / A
```

`A` and `Nv` measured from the built models (meta device):

| flavor | A (active) | Nv (vision) | r=4Nv/A | err @ f=0.1 | err @ f=0.5 | err @ f=0.9 |
|---|---|---|---|---|---|---|
| 0.8B      | 0.75B  | 101M | 0.535 | +7.6%  | -10.6% | -23.5% |
| 2B        | 1.88B  | 331M | 0.704 | +9.9%  | -13.0% | -28.0% |
| 4B        | 4.21B  | 334M | 0.317 | +4.6%  | -6.8%  | -16.0% |
| 9B        | 7.94B  | 456M | 0.230 | +3.4%  | -5.2%  | -12.4% |
| 27B       | 25.6B  | 461M | 0.072 | +1.1%  | -1.7%  | -4.4%  |
| 35B-A3B   | 2.95B  | 447M | 0.606 | +8.6%  | -11.6% | -25.5% |
| 122B-A10B | 9.01B  | 451M | 0.200 | +2.9%  | -4.6%  | -11.0% |
| 397B-A17B | 16.3B  | 456M | 0.112 | +1.7%  | -2.6%  | -6.6%  |

(+ = MFU overstated, - = understated.) The error is dominant at the small end -- up to ~28% for the 2B rung and ~25% for the small MoE rung with image-heavy inputs -- and negligible by 27B. It is governed by *active* params, so small-active MoE rungs are hit as hard as the small dense ones.

### Why it is not a one-liner

`num_flops_per_token` is a static per-token scalar consumed by the metrics/MFU path. Vision FLOPs are **data-dependent** -- they vary per step with the number of images and their resolution (`grid_thw`) -- so they do not fit a fixed per-token constant. A correct fix needs a per-step vision-FLOP contribution derived from the actual patch counts, added to the step FLOP total before dividing by time.

### Proposed direction

1. Exclude `vision_encoder` params from `nparams_for_matmul` so they are no longer billed per decoder token.
2. Add a per-step vision-encoder FLOP term computed from the batch's grid sizes (`~6 * Nv * num_patches`, plus the ViT attention `O(L^2)` term) to the step FLOP total in the MFU accounting.
3. Until then, document that VLM MFU is an approximation and note the direction/magnitude of the error.

### PR DESCRIPTION
**Pull request by jinsooihm**
Resolves #4009 

Previously, each models returned a function `get_nparams_and_flops` which returned `num_flops_per_token`, which the trainer used to compute the FLOPs by multiplying the number of text tokens seen to it. 

This does not take into account of FLOPs of multimodal models. The computation becomes more complicated, because vision inputs may have images of different sizes. 

## Change
Each model now returns a function `FlopsEstimator` that takes in a input batch and returns the FLOPs for that input batch. FLOPs is computed and accumulated locally per input batch in the CPU, and it is gathered at the logging step.

Decoder only model FLOPs unchanged.
| Model / input | n | Main: FLOPs/token × n | New estimator | Change |
|---|---:|---:|---:|---:|
| Llama 3 text | 16 | 546,103,296 | 546,103,296 | 0.0000% |
| Qwen 3 MoE text | 16 | 4,945,698,816 | 4,945,698,816 | 0.0000% |
| GPT-OSS text | 16 | 4,416,359,424 | 4,416,359,424 | 0.0000% |
| DeepSeek V3 + MTP text | 8 | 1,586,823,168 | 1,586,823,168 | 0.0000% |
| DeepSeek V4 Flash + MTP text | 4,096 | 379,954,597,380,096 | 379,954,597,380,096 | 0.0000% |
| HF backend dense text | 16 | 5,583,839,232 | 5,583,839,232 | 0.0000% |
| HF backend MoE text | 16 | 35,340,976,128 | 35,340,976,128 | 0.0000% |
| Qwen 3.5/3.6/3.8, image `(1,4,4)` + video `(2,4,4)` | 16 | 6,565,701,888 | 7,081,963,776 | +7.8630% |
| Kimi K2.7, image `(1,4,4)` | 16 | 2,694,340,608 | 2,945,955,840 | +9.3387% |
| Kimi K2.7, video `(2,4,4)` | 16 | 2,694,340,608 | 3,172,325,376 | +17.7403% |
| Kimi K3, image `(1,4,4)` | 16 | 52,826,950,656 | 55,845,153,792 | +5.7134% |
| Muse Glimmer, image `(1,4,4)` | 16 | 711,745,536 | 731,043,840 | +2.7114% |
| Muse Glimmer, images `(1,4,4)` + `(1,6,8)` | 16 | 711,745,536 | 791,298,048 | +11.1771% |
| Flux, batch 2 at 256×256 (`seq_len=512`) | 1,024 | 814,721,925,120 | 814,721,925,120 | 0.0000% |

## Limitations
This PR makes the FLOPs calculation take into account of each images in the batch, but it does not make the calculation aware of the different document lengths present in an input text sequence.

### CODE REVIEW

**Review by SherlockNoMad — APPROVED** [review]
graph trainer change looks fine to me

File: torchtitan/models/utils.py [hunk 1]
```diff
@@ -503,19 +499,37 @@ def get_nparams_and_active_nparams(
...
-    active_nparams = nparams_for_matmul.numerator
+    assert usage_weighted_nparams.denominator == 1
+    return 6 * usage_weighted_nparams.numerator
 
-    logger.info(
-        f"Total parameter count: {nparams:,}, active parameters: {active_nparams:,}"
-    )
-    return nparams, active_nparams
+
+def build_token_flops_estimator(
```
Comment (tianyu-l) [inline torchtitan/models/utils.py:515]: already having `build_flops_estimator`, so prefer to not have this thin wrapper with similar name

File: torchtitan/models/utils.py [hunk 2]
```diff
@@ -503,19 +499,37 @@ def get_nparams_and_active_nparams(
...
-        param.numel() * parameter_weights[id(param)] for _, param in named_parameters
+    usage_weighted_nparams = sum(
+        param.numel() * parameter_weights[id(param)] for param in parameters
     )
-    assert nparams_for_matmul.denominator == 1
-    active_nparams = nparams_for_matmul.numerator
+    assert usage_weighted_nparams.denominator == 1
+    return 6 * usage_weighted_nparams.numerator
 
-    logger.info(
```
Comment (tianyu-l) [inline torchtitan/models/utils.py:518]: is this gone from CLI?

File: torchtitan/training_engine.py [hunk 3]
```diff
@@ -287,11 +284,10 @@ def _initialize_model(
         ):
             model = self.model_config.build()
         model.verify_module_protocol()
-        (
-            self.model_param_count,
-            self.num_flops_per_token,
-        ) = self.model_config.get_nparams_and_flops(
-            model, self.config.training.max_context_length
+        self.model_param_count = sum(param.numel() for param in model.parameters())
```
Comment (tianyu-l) [inline torchtitan/training_engine.py:287]: but I'm losing active param count

File: torchtitan/models/utils.py [hunk 4]
```diff
@@ -503,19 +499,37 @@ def get_nparams_and_active_nparams(
...
-    return nparams, active_nparams
+
+def build_token_flops_estimator(
+    flops_per_token: int,
+) -> Callable[[Mapping[str, Any]], int]:
+    """Build an estimator for work that scales with input token count."""
+    return lambda batch: flops_per_token * batch["input"].numel()
+
+
+def sum_packed_grid_flops(
```
Comment (tianyu-l) [inline torchtitan/models/utils.py:522]: should this go to multimodal.py?

File: torchtitan/models/flux/model/model.py [hunk 5]
```diff
@@ -77,9 +78,20 @@ def update_from_config(self, *, config, **kwargs) -> None:
...
-        def get_nparams_and_flops(
-            self, model: nn.Module, seq_len: int
-        ) -> tuple[int, int]:
+        def build_flops_estimator(
+            self,
+            model: nn.Module,
+            *,
+            seq_len: int,
+        ) -> FlopsEstimator:
+            cached_flops_per_token = self._estimate_flops_per_token(model, seq_len)
```
Comment (tianyu-l) [inline torchtitan/models/flux/model/model.py:87]: what is "cached" referring to here?

File: torchtitan/observability/metrics.py [hunk 6]
```diff
@@ -392,10 +395,15 @@ def __init__(
         self.has_quantization = has_quantization
 
         # These variables have to be set later as they depend on other components or model.
-        self.num_flops_per_token = -1
         self.optimizers = None
         self.model_parts = None
 
+    def record_step_flops(self, num_flops: int) -> None:
```
Comment (tianyu-l) [inline torchtitan/observability/metrics.py:401]: "step" is ambiguous

File: torchtitan/trainer.py [hunk 7]
```diff
@@ -363,6 +360,12 @@ def train_step(self, data_iterator: Iterator[TrainingMicrobatch]):
                 local_valid_tokens += microbatch.num_valid_tokens
                 microbatch_group.append(microbatch)
             microbatch_groups.append(microbatch_group)
+
+        step_num_flops = sum(
```
Comment (tianyu-l) [inline torchtitan/trainer.py:364]: step is ambiguous -- could be optimizer step / fwd_bwd_microbatch step, etc.

File: torchtitan/trainer.py [hunk 8]
```diff
@@ -363,6 +360,12 @@ def train_step(self, data_iterator: Iterator[TrainingMicrobatch]):
...
+            engine.estimate_flops(microbatch.as_input_dict())
```
Comment (tianyu-l) [inline torchtitan/trainer.py:365]: need to be very careful about D2H / H2D moves, to not create additional sync point between GPU and CPU

File: torchtitan/models/utils.py [hunk 9]
```diff
@@ -503,19 +499,37 @@ def get_nparams_and_active_nparams(
...
-    logger.info(
-        f"Total parameter count: {nparams:,}, active parameters: {active_nparams:,}"
-    )
-    return nparams, active_nparams
+
+def build_token_flops_estimator(
+    flops_per_token: int,
+) -> Callable[[Mapping[str, Any]], int]:
+    """Build an estimator for work that scales with input token count."""
+    return lambda batch: flops_per_token * batch["input"].numel()
```
Comment (tianyu-l) [inline torchtitan/models/utils.py:519]: "batch" is not great name, because sometimes you call this function on a microbatch

File: torchtitan/distributed/utils.py [hunk 10]
```diff
@@ -104,6 +104,52 @@ def dist_sum_tensor(
...
+        dist_sum_tensor(
+            local_num_flops_tensor,
+            mesh=mesh,
+            extra_pg=extra_pg,
+        )
+        / divisor
+    )
+
+
+def materialize_scalar_tensors(
```
Comment (tianyu-l) [inline torchtitan/distributed/utils.py:131]: maybe is it not worth the complexity? if only concatenating two tensors in log steps only

File: torchtitan/distributed/utils.py [hunk 11]
```diff
@@ -104,6 +104,52 @@ def dist_sum_tensor(
     )
 
 
+def mean_flops_tensor(
```
Comment (tianyu-l) [inline torchtitan/distributed/utils.py:107]: prefer not having this thin wrappers

File: torchtitan/rl/trainer.py [hunk 12]
```diff
@@ -309,10 +312,25 @@ async def optimizer_step(self, *, last_step: bool = False) -> OptimizerStepOutpu
...
+        grad_norm_value, mean_num_flops = dist_utils.materialize_scalar_tensors(
+            [grad_norm, self._step_mean_num_flops_tensor]
+        )
+
+        # TODO: Move performance, LR, and auxiliary-loss reporting into a shared
+        # trainer metrics interface while preserving controller-side aggregation.
+        performance = compute_training_performance_metrics(
+            num_tokens=self._step_num_tokens_per_dp_rank,
+            elapsed_time=elapsed_time,
+            non_data_parallel_size=engine.parallel_dims.non_data_parallel_size,
```
Comment (tianyu-l) [inline torchtitan/rl/trainer.py:324]: when computing flogs, this flag should be used with caution, because we have CP replicate input -> CP shard -> loss mesh reduction

File: torchtitan/models/kimi_k2_7/model.py [hunk 13]
```diff
@@ -88,16 +92,74 @@ def update_from_config(
...
+        ) -> tuple[nn.Module | None, ...]:
             kimi_model = cast("KimiK25Model", model)
-            return get_kimi_k2_7_nparams_and_flops(
-                self,
-                model,
-                seq_len,
-                modules_excluded_from_active_params=(kimi_model.vision_encoder,),
+            return (kimi_model.vision_encoder,)
+
+        def build_flops_estimator(
```
Comment (tianyu-l) [inline torchtitan/models/kimi_k2_7/model.py:101]: maybe only multimodal defines it once and each encoder is responsible for their own specialization; maybe same for attention variants (maybe not differentiate recurrent vs. chunk for linear attention)

File: torchtitan/models/utils.py (hunk 2 above)
Comment (jinsooihm) [reply torchtitan/models/utils.py:518]: Should the vision encoder parameters also be included in the logged `num_active_parameters`? Previously, it was not included.

File: torchtitan/observability/metrics.py (hunk 6 above)
Comment (jinsooihm) [reply torchtitan/observability/metrics.py:401]: I will change to `optimizer_step`

File: torchtitan/models/utils.py (hunk 2 above)
Comment (tianyu-l) [reply torchtitan/models/utils.py:518]: I feel we should, but please check with popular multimodal models -- when they publish a model, claim the size, do they count the vision encoder?

Comment (jinsooihm) [reply torchtitan/models/utils.py:518]: ## takeaways
- vision encoder is usually very small, so it is hard to say whether it was included in the param count or active param count after rounding ( I guess more often not )
- published Qwen 3.5 numbers seems to be after weight tying, so calling literal sum of `param.numel()` gives wrong number
- Torchtitan excludes embedding from active param
    - qwen 3.5 includes embedding in active param
    - kimi seems to exclude it


| Family | Flavor | Published total | Published active | TorchTitan total | Vision encoder | TT decoder-active | TT active + encoder |
|---|---|---:|---:|---:|---:|---:|---:|
| Kimi | Kimi-VL-A3B | ~16B | ~3B | 16.408B | 447.546M | 2.579B | 3.027B |
| Kimi | Kimi-K2.5 / K2.6 / K2.7-Code | ~1T | ~32B | 1,026.879B | 471.144M | 31.687B | 32.158B |
| Kimi | Kimi-K3 | 2.8T | 104B | 2,779.932B | 447.359M | 104.190B | 104.637B |
| Muse Glimmer | 30B_mm | 30B | Not separately stated | 29.777B | 1,852.640M | 26.510B | 28.363B |
| Qwen3.5 | 0.8B | 0.8B | Not separately stated (dense) | 1.107B | 100.593M | 0.752B | 0.853B |
| Qwen3.5 | 2B | 2B | Not separately stated (dense) | 2.722B | 331.417M | 1.882B | 2.213B |
| Qwen3.5 | 4B | 4B | Not separately stated (dense) | 5.175B | 333.514M | 4.206B | 4.539B |
| Qwen3.5 | 9B | 9B | Not separately stated (dense) | 9.410B | 456.010M | 7.937B | 8.393B |
| Qwen3.5 / 3.6 / 3.8 | 27B | 27B | Not separately stated (dense) | 27.357B | 460.730M | 25.625B | 26.085B |
| Qwen3.5 / 3.6 | 35B-A3B | 35B | 3B | 35.107B | 446.571M | 2.946B | 3.393B |
| Qwen3.5 | 122B-A10B | 122B | 10B | 122.563B | 451.291M | 9.008B | 9.460B |
| Qwen3.5 | 397B-A17B | 397B | 17B | 396.802B | 456.010M | 16.332B | 16.788B |

Comment (jinsooihm) [reply torchtitan/models/utils.py:518]: There seems to be no one nice way to exactly match the official published param/active param count. 

I think for our numbers we should make a choice:
- total_parameters: every unique parameter in the actual torchtitan model, no double counting weight tied params. 
- active_parameters: MoE consideration + embedding params + multimodal encoders

And params used for matmul flops would exclude the embedding params, so it will not directly use active_parameters.

File: torchtitan/models/flux/model/model.py (hunk 5 above)
Comment (jinsooihm) [reply torchtitan/models/flux/model/model.py]: Cacheing for capturing by the closure below. I will just change it to `flops_per_token`.
